python/lib/modeldata/dmod/modeldata/data/bmi_init_config_auto_generator.py

```diff
@@ -0,0 +1,475 @@
+        # https://github.com/NOAA-OWP/noah-owp-modular/blob/30d0f53e8c14acc4ce74018e06ff7c9410ecc13c/docs/changelog.md?plain=1#L35C7-L35C168
+        # > SOILCOLOR is hard-coded as 4 in module_sf_noahmpdrv.F in the current release of HRLDAS. SOILCOLOR is used to select the albedo values for dry and saturated soil.
+        # NOTE: it appears that the soil color indexes into the ALBSAT_VIS, ALBSAT_NIR, ALBDRY_VIS, ALBDRY_NIR tables
+        # https://github.com/NOAA-OWP/noah-owp-modular/blob/30d0f53e8c14acc4ce74018e06ff7c9410ecc13c/parameters/MPTABLE.TBL#L328-L331
+        # here is the indexing
+        # https://github.com/NOAA-OWP/noah-owp-modular/blob/30d0f53e8c14acc4ce74018e06ff7c9410ecc13c/src/ParametersType.f90#L303-L306
+        # NOTE: looks like for HRLDAS this is 4
+        soilcolor: int = 4
+        structure = Structure(
+            isltyp=isltyp,
+            nsoil=nsoil,
+            nsnow=nsnow,
+            nveg=nveg,
+            vegtyp=vegtyp,
+            # crop type (SET TO 0, no crops currently supported)
+            # source: https://github.com/NOAA-OWP/noah-owp-modular/blob/30d0f53e8c14acc4ce74018e06ff7c9410ecc13c/src/NamelistRead.f90#L36
+            croptype=croptype,
+            sfctyp=sfctyp,
+            soilcolor=soilcolor,
+        )
+        self.data["structure"] = structure
+
+    def visit(self, hook_provider: "HookProvider") -> None:
+        hook_provider.provide_hydrofabric_linked_data(self)
+
+        self._v2_defaults()
+
+    def build(self) -> BaseModel:
+        return NoahOWPConfig(**self.data)
+
+
+class BmiGenDivideIdHookObject:
+    def __init__(self):
+        self.__divide_id: Union[str, None] = None
+
+    def hydrofabric_hook(self, version: str, divide_id: str, data: Dict[str, Any]) -> None:
+        self.__divide_id = divide_id
+
+    def visit(self, hook_provider) -> None:
+        hook_provider.provide_hydrofabric_data(self)
+
+    def divide_id(self) -> Union[str, None]:
+        return self.__divide_id
+
+
+# TODO: figure out how to handle noah owp
+class BmiInitConfigAutoGenerator:
+    """
+    A tool for automatically generating required BMI initialization configs for a realization config and hydrofabric.
+
+    A helper tool for generating BMI init configs as needed for a given realization config.  It also requires data from
+    a hydrofabric dataset, not only because a realization config only has real meaning in the context of a particular
+    hydrofabric (i.e., otherwise, what exactly is, for example, ``cat-115`` referring to?), but because the hydrofabric
+    dataset contains model attributes data from which an applicable BMI init config can be generated.
+
+    This type strategically applies classes in the ``ngen.config_gen`` package to perform the actual generation of
+    configs.
+
+    See https://www.lynker-spatial.com/copyright.html for license details on hydrofabric data.
+    """
+
+    _module_to_model_map: Dict[str, Any] = {"CFE": Cfe, "PET": Pet}
```

Comment:
What are your thoughts on moving this out of the class and passing this in as a parameter to `__init__`? My thinking is this will provide a smoother pathway for adding support for new modules as they are introduced.

Reply:
Hmm, not sure.  I see the benefit of parameterizing things; I added the `other_builder_hook_types` init param to future proof along those lines.  But I don't see any reason why we ever wouldn't want these available to an instance, certainly for the moment but probably also in the long term.  So they should be hard-coded somewhere.  And where is better than here?

Reply:
Yeah that is a fair point. My thinking was that the code that uses this would parametrize it. So, that would most likely be in a service instead of a library. My thinking is it will be easier / less painful to update service code rather than library code.

Reply:
> My thinking is it will be easier / less painful to update service code rather than library code.

Maybe, but only if we only need to do that once.  If we are putting _BmiInitConfigAutoGenerator_ into a library, we intend it to at least eventually receive usage in more than just one place in one service (it's a fair question as to whether we _should_ do this or not, but for the moment we weren't asking that). 

Again though, I assume for this that an unrecognized `model_type_name` in the realization config should be treated as invalid, from which follows that certain builders should always be available to an instance.